### PR TITLE
Benchtool: Separate command generation and submission. Also other clean-ups [DPP-978]

### PR DESCRIPTION
**Pull request by pbatko-da**
This is setting ground for generating contracts with non-empty non-stakeholder informees in https://github.com/digital-asset/daml/pull/13808

changelog_begin
changelog_end

### CODE REVIEW

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+          createEvents.addOne(
+            ObservedCreateEvent(
+              templateName = templateName,
+              createArgumentsSerializedSize = argsSize,
+            )
+          )
+        }
+      }
+    }
+  }
```
Comment (skisel-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala:48]:
```suggestion
  override def onNext(value: GetTransactionsResponse): Unit =
    for {
      transaction <- value.transactions
      event <- transaction.events
      created <- event.event.created.toList
    } {
      val argsSize = created.createArguments.fold(0)(_.serializedSize)
      val templateName =
        created.templateId.getOrElse(sys.error(s"Expected templateId in $created")).entityName
      val observedCreateEvent = ObservedCreateEvent(
        templateName = templateName,
        createArgumentsSerializedSize = argsSize,
      )
      createEvents.addOne(observedCreateEvent)
    }
```

As I am known hater of `} } } } } } }` ;)

Comment (skisel-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala:48]:
```suggestion
  override def onNext(value: GetTransactionsResponse): Unit =
    for {
      transaction <- value.transactions
      event <- transaction.events
      created <- event.event.created.toList
    } {
      val argsSize = created.createArguments.fold(0)(_.serializedSize)
      val templateName =
        created.templateId.getOrElse(sys.error(s"Expected templateId in $created")).entityName
      val observedCreateEvent = ObservedCreateEvent(
        templateName = templateName,
        createArgumentsSerializedSize = argsSize,
      )
      createEvents.addOne(observedCreateEvent)
    }
```

As I am known hater of `} } } } } } }` ;)

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/TreeEventsObserver.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+              templateName = templateName,
+              choiceName = choiceName,
+              choiceArgumentsSerializedSize = argsSize,
+              consuming = exercised.consuming,
+            )
+          )
+        }
+      }
+    }
+  }
```
Comment (skisel-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/TreeEventsObserver.scala:119]:
```suggestion
  override def onNext(value: GetTransactionTreesResponse): Unit = {
    for {
      transaction <- value.transactions
      allEvents = transaction.eventsById.values
      event <- allEvents
    } yield {
      event.kind.created.foreach(created => createEvents.addOne(ObservedCreateEvent(created)))
      event.kind.exercised.foreach(exercised =>
        exerciseEvents.addOne(ObservedExerciseEvent(exercised))
      )
    }
  }
```

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/TreeEventsObserver.scala [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import scala.concurrent.Future
+
+case class ObservedExerciseEvent(
+    templateName: String,
+    choiceName: String,
+    choiceArgumentsSerializedSize: Int,
+    consuming: Boolean,
+)
+
+case class ObservedCreateEvent(templateName: String, createArgumentsSerializedSize: Int)
```
Comment (skisel-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/TreeEventsObserver.scala:21]:
```suggestion
case class ObservedExerciseEvent(
    templateName: String,
    choiceName: String,
    choiceArgumentsSerializedSize: Int,
    consuming: Boolean,
)
object ObservedExerciseEvent {
  def apply(exercised: com.daml.ledger.api.v1.event.ExercisedEvent): ObservedExerciseEvent = {
    val argsSize = exercised.choiceArgument.fold(0)(_.serializedSize)
    val templateName = exercised.templateId
      .getOrElse(sys.error(s"Expected templateId in $exercised"))
      .entityName
    val choiceName = exercised.choice
    ObservedExerciseEvent(
      templateName = templateName,
      choiceName = choiceName,
      choiceArgumentsSerializedSize = argsSize,
      consuming = exercised.consuming,
    )
  }
}

case class ObservedCreateEvent(templateName: String, createArgumentsSerializedSize: Int)
object ObservedCreateEvent {
  def apply(created: com.daml.ledger.api.v1.event.CreatedEvent): ObservedCreateEvent = {
    val argsSize = created.createArguments.fold(0)(_.serializedSize)
    val templateName =
      created.templateId.getOrElse(sys.error(s"Expected templateId in $created")).entityName
    ObservedCreateEvent(
      templateName = templateName,
      createArgumentsSerializedSize = argsSize,
    )
  }
}
```

**Review by skisel-da — APPROVED** [review]

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala [hunk 4]
```diff
@@ -207,39 +118,214 @@ object LedgerApiBenchTool {
...
+                regularUserServices = regularUserServices,
+                adminServices = adminServices,
+                submissionConfig = submissionConfig,
+                metricRegistry = metricRegistry,
+              ).map(Some(_))
+          }
+        }
+
+        updatedStreamConfigs = config.workflow.streams.map(streamsConfig =>
+          ConfigEnricher.enrichStreamConfig(streamsConfig, submissionStepResultO)
```
Comment (simonmaxen-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala:137]: This seems more work than passing in the submissionStepResult below.

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      created <- event.event.created.toList
+    } {
+      val argsSize = created.createArguments.fold(0)(_.serializedSize)
+      val templateName =
+        created.templateId.getOrElse(sys.error(s"Expected templateId in $created")).entityName
+      val observedCreateEvent = ObservedCreateEvent(
+        templateName = templateName,
+        createArgumentsSerializedSize = argsSize,
+      )
+      createEvents.addOne(observedCreateEvent)
```
Comment (skisel-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala:42]:
```suggestion
      createEvents.addOne(ObservedCreateEvent(created))
```

Comment (skisel-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala:42]: sorry I suggested this specific line before I introduced another constructor in `ObservedCreateEvent`

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala [hunk 6]
```diff
@@ -207,39 +118,214 @@ object LedgerApiBenchTool {
...
         )
+
         _ = logger.info(
-          s"Stream configs adapted after the submission step: ${prettyPrint(streams)}"
+          s"Stream configs adapted after the submission step: ${prettyPrint(updatedStreamConfigs)}"
         )
         benchmarkResult <-
           if (config.latencyTest) {
```
Comment (simonmaxen-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala:144]: When I first read this it seemed to me we had the wrong abstraction.

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/CommandGenerator.scala [hunk 7]
```diff
@@ -3,30 +3,10 @@
...
-import com.daml.ledger.client.binding.Primitive
 
 import scala.util.Try
 
 trait CommandGenerator {
   def next(): Try[Seq[Command]]
 }
-
-object CommandGenerator {
-  def apply(
```
Comment (simonmaxen-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/CommandGenerator.scala:21]: This is an extension point for additional generators, in-lining this logic means that to add another generator requires editing and extending `LedgerApiBenchTool` which is possibly already too long and complicated.

**Review by simonmaxen-da — APPROVED** [review]

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala (hunk 4 above)
Comment (pbatko-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala:137]: Discussed offline

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala (hunk 6 above)
Comment (pbatko-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala:144]: Discussed offline

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala [hunk 8]
```diff
@@ -13,7 +12,7 @@ object ConfigEnricher {
 
   def enrichStreamConfig(
       streamConfig: StreamConfig,
-      submissionSummary: Option[CommandSubmitter.SubmissionSummary],
+      submissionSummary: Option[SubmissionStepResult],
```
Comment (mziolekda) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala:15]: You have changed the type, but that also changes the semantic of the field. Shouldn't it be reflected in the parameter name?

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+      created <- event.event.created.toList
+    } {
+      createEvents.addOne(ObservedCreateEvent(created))
+    }
+
+  override def completeWith(): Future[ObservedEvents] = Future.successful(
+    ObservedEvents(
+      expectedTemplateNames = expectedTemplateNames,
+      createEvents = createEvents.toList,
+      exerciseEvents = exerciseEvents.toList,
```
Comment (mziolekda) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala:42]: That will always be 0, won't it?

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+/** Collects information about create and exercise events.
+  */
+class FlatEventsObserver(expectedTemplateNames: Set[String], logger: Logger)
+    extends ObserverWithResult[GetTransactionsResponse, ObservedEvents](logger) {
+
+  private val createEvents = collection.mutable.ArrayBuffer[ObservedCreateEvent]()
+  private val exerciseEvents = collection.mutable.ArrayBuffer[ObservedExerciseEvent]()
+
+  override def streamName: String = "dummy-stream-name"
```
Comment (tudor-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala:27]: Is this a used in unit test or is this dummy override an oversight?

Comment (pbatko-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala:27]: That's used just in tests so I think it's fine. 
I also plan to move these observers to a test-lib target

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala (hunk 9 above)
Comment (pbatko-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/FlatEventsObserver.scala:42]: For non-stakeholders 0, for stakeholders not necessarily

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala (hunk 8 above)
Comment (pbatko-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala]: Renamed

**Pull request merged**